[CI] BlobStoreSyncDirectoryTests testCloseCancelsOnGoingUploads failing

**Build Scans:**
- [elasticsearch-intake #41798 / part2](https://gradle-enterprise.elastic.co/s/nyim5rvl4q3cq)
- [elasticsearch-pull-request #155396 / part-2](https://gradle-enterprise.elastic.co/s/un4xzjarggbwg)
- [elasticsearch-pull-request #155371 / part-2](https://gradle-enterprise.elastic.co/s/2fnyedwuijvsa)
- [elasticsearch-pull-request #155357 / part-2](https://gradle-enterprise.elastic.co/s/dyaocftwfvbyk)
- [elasticsearch-intake #41676 / part2](https://gradle-enterprise.elastic.co/s/5aurkimydchvq)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:stateless:test" --tests "org.elasticsearch.xpack.stateless.cluster.coordination.BlobStoreSyncDirectoryTests.testCloseCancelsOnGoingUploads" -Dtests.seed=E91D3B37F845E5DE -Dtests.locale=kkj-CM -Dtests.timezone=Africa/Ndjamena -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.stateless.cluster.coordination.BlobStoreSyncDirectoryTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testCloseCancelsOnGoingUploads'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: expected java.lang.Exception to be thrown, but nothing was thrown
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testCloseCancelsOnGoingUploads
- [main] 1 consecutive failures in step part2
- [main] 1 consecutive failures in pipeline elasticsearch-intake
- [main] 5 failures in test testCloseCancelsOnGoingUploads (3.7% fail rate in 134 executions)
- [main] 2 failures in step part2 (7.4% fail rate in 27 executions)
- [main] 3 failures in step part-2 (3.0% fail rate in 99 executions)
- [main] 2 failures in pipeline elasticsearch-intake (7.4% fail rate in 27 executions)
- [main] 3 failures in pipeline elasticsearch-pull-request (3.1% fail rate in 97 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -626,9 +626,6 @@ tests:
 - class: org.elasticsearch.xpack.stateless.StatelessMergePreWarmingIT
   method: testMergePreWarmingFailureDoesNotFailTheEngine
   issue: https://github.com/elastic/elasticsearch/issues/151618
-- class: org.elasticsearch.xpack.stateless.cluster.coordination.BlobStoreSyncDirectoryTests
-  method: testCloseCancelsOnGoingUploads
-  issue: https://github.com/elastic/elasticsearch/issues/151622
 - class: org.elasticsearch.gradle.internal.precommit.PomValidationPrecommitPluginFuncTest
   method: detects missing POM elements and reports structured problems
   issue: https://github.com/elastic/elasticsearch/issues/151636

--- a/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cluster/coordination/BlobStoreSyncDirectoryTests.java
+++ b/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cluster/coordination/BlobStoreSyncDirectoryTests.java
@@ -278,18 +278,12 @@ public class BlobStoreSyncDirectoryTests extends ESTestCase {
                         long blobSize,
                         boolean failIfAlreadyExists
                     ) throws IOException {
+                        uploadedFiles.add(blobName);
                         super.writeBlob(purpose, blobName, inputStream, blobSize, failIfAlreadyExists);
                     }
                 };
             }
-        };
-            var dir = new BlobStoreSyncDirectory(
-                new ByteBuffersDirectory(),
-                () -> statelessNode.objectStoreService.getClusterStateBlobContainerForTerm(1),
-                statelessNode.threadPool.executor(ThreadPool.Names.SNAPSHOT)
-            );
-            var indexWriter = new IndexWriter(dir, getIndexWriterConfig())
-        ) {
+        }) {
             var threadPool = statelessNode.threadPool;
             var snapshotExecutor = (EsThreadPoolExecutor) threadPool.executor(ThreadPool.Names.SNAPSHOT);
             var maxPoolSize = snapshotExecutor.getMaximumPoolSize();
@@ -306,22 +300,43 @@ public class BlobStoreSyncDirectoryTests extends ESTestCase {
                 });
             }
 
-            indexRandomDocuments(indexWriter, randomIntBetween(100, 200));
-            var commitFuture = threadPool.executor(ThreadPool.Names.GENERIC).submit(() -> {
+            var bytesDirectory = new ByteBuffersDirectory();
+            var filesToUpload = createRandomFiles(bytesDirectory, randomIntBetween(10, 100));
+            var dir = new BlobStoreSyncDirectory(
+                bytesDirectory,
+                () -> statelessNode.objectStoreService.getClusterStateBlobContainerForTerm(1),
+                snapshotExecutor
+            );
+
+            var genericExecutor = threadPool.executor(ThreadPool.Names.GENERIC);
+            var syncFuture = genericExecutor.submit(() -> {
                 try {
-                    indexWriter.commit();
+                    dir.sync(filesToUpload);
                 } catch (IOException e) {
                     throw new RuntimeException(e);
                 }
             });
-
             dir.close();
-
             blockExecutionLatch.countDown();
 
-            expectThrows(Exception.class, commitFuture::get);
+            try {
+                syncFuture.get();
+            } catch (Exception ignored) {}
+
             assertThat(uploadedFiles, is(empty()));
         }
+    }
+
+    private static List<String> createRandomFiles(ByteBuffersDirectory directory, int numberOfDocs) throws IOException {
+        List<String> files = new ArrayList<>();
+        for (int i = 0; i < numberOfDocs; i++) {
+            final var fileName = "segments_" + i;
+            try (var out = directory.createOutput(fileName, IOContext.DEFAULT)) {
+                out.writeString(randomAlphaOfLengthBetween(10, 100));
+                files.add(fileName);
+            }
+        }
+        return files;
     }
 
     private void indexRandomDocuments(IndexWriter indexWriter, int numberOfDocsSecondCommit) throws IOException {

--- a/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cluster/coordination/BlobStoreSyncDirectoryTests.java
+++ b/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cluster/coordination/BlobStoreSyncDirectoryTests.java
@@ -321,7 +321,9 @@ public class BlobStoreSyncDirectoryTests extends ESTestCase {
 
             try {
                 syncFuture.get();
-            } catch (Exception ignored) {}
+            } catch (Exception ignored) {
+                fail("we don't expect any exception to be thrown during sync");
+            }
 
             assertThat(uploadedFiles, is(empty()));
         }

--- a/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cluster/coordination/BlobStoreSyncDirectoryTests.java
+++ b/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/cluster/coordination/BlobStoreSyncDirectoryTests.java
@@ -319,12 +319,7 @@ public class BlobStoreSyncDirectoryTests extends ESTestCase {
             dir.close();
             blockExecutionLatch.countDown();
 
-            try {
-                syncFuture.get();
-            } catch (Exception ignored) {
-                fail("we don't expect any exception to be thrown during sync");
-            }
-
+            assertThrows(Exception.class, syncFuture::get);
             assertThat(uploadedFiles, is(empty()));
         }
     }

